### PR TITLE
Run pyupgrade for python 3.7 & 3.8.

### PR DESCRIPTION
**Pull request by clokep**
Two commits running `pyupgrade --py37-plus --keep-percent-format **/*.py`, then `pyupgrade --py38-plus --keep-percent-format **/*.py`.

(Note that ruff seems to now support using pyupgrade, but I had issues enabling it.)

### CODE REVIEW

File: synapse/handlers/sso.py [hunk 1]
```diff
@@ -24,13 +24,14 @@
...
     Mapping,
+    NoReturn,
     Optional,
     Set,
 )
 from urllib.parse import urlencode
 
 import attr
-from typing_extensions import NoReturn, Protocol
+from typing_extensions import Protocol
```
Comment (DMRobertson) [inline synapse/handlers/sso.py:34]: Protocol was added in 3.8, see https://docs.python.org/3/library/typing.html?highlight=typing%20protocol#typing.Protocol, so wecan import this from `typing` now.

File: synapse/handlers/stats.py [hunk 2]
```diff
@@ -14,9 +14,15 @@
 # limitations under the License.
 import logging
 from collections import Counter
-from typing import TYPE_CHECKING, Any, Dict, Iterable, Optional, Tuple
-
-from typing_extensions import Counter as CounterType
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    Counter as CounterType,
```
Comment (DMRobertson) [inline synapse/handlers/stats.py:20]: Looks like we should be able to just use `collections.Counter[str]` in 3.9: https://docs.python.org/3/library/typing.html?highlight=typing%20protocol#typing.Counter

File: synapse/storage/databases/main/keys.py [hunk 3]
```diff
@@ -188,7 +188,7 @@ async def store_server_keys_json(
         # invalidate takes a tuple corresponding to the params of
         # _get_server_keys_json. _get_server_keys_json only takes one
         # param, which is itself the 2-tuple (server_name, key_id).
-        self._get_server_keys_json.invalidate((((server_name, key_id),)))
+        self._get_server_keys_json.invalidate(((server_name, key_id),))
```
Comment (DMRobertson) [inline synapse/storage/databases/main/keys.py:191]: I think this is safe: before we have `invalidate( (t1) )` where `t1 = (t2,)`. We are removing the inner brackets around `t1`.

File: synapse/types/__init__.py [hunk 4]
```diff
@@ -38,7 +39,7 @@
 from immutabledict import immutabledict
 from signedjson.key import decode_verify_key_bytes
 from signedjson.types import VerifyKey
-from typing_extensions import Final, TypedDict
+from typing_extensions import TypedDict
```
Comment (DMRobertson) [inline synapse/types/__init__.py:42]: FWIW TypedDict was new in https://docs.python.org/3/library/typing.html?highlight=typeddict#typing.TypedDict , though there are various changes that were made in later Python releases

File: tests/replication/test_multi_media_repo.py [hunk 5]
```diff
@@ -116,7 +116,7 @@ def _get_media_req(
         self.assertEqual(request.method, b"GET")
         self.assertEqual(
             request.path,
-            f"/_matrix/media/r0/download/{target}/{media_id}".encode("utf-8"),
+            f"/_matrix/media/r0/download/{target}/{media_id}".encode(),
```
Comment (DMRobertson) [inline tests/replication/test_multi_media_repo.py:119]: Presumably `utf-8` is the default?

**Review by DMRobertson — APPROVED** [review]

File: synapse/handlers/sso.py (hunk 1 above)
Comment (DMRobertson) [reply synapse/handlers/sso.py:34]: (I think ignore this comment and merge this as-is. We'll still need typing extensions for things like ParamSpec.)

Comment (clokep) [reply synapse/handlers/sso.py:34]: I'm surprised that it didn't do this automatically TBH...

File: synapse/storage/databases/main/keys.py (hunk 3 above)
Comment (clokep) [reply synapse/storage/databases/main/keys.py:191]: Yeah, we have 4 sets of parens, if you expand it out you have:

1. Function call
2. Grouping (unneeded)
3. Tuple
4. Tuple

```python
invalidate( #function call
    ( # grouping
        ( # tuple
            ( # tuple
                server_name, key_id
            ),
        )
    )
)
```

File: tests/replication/test_multi_media_repo.py (hunk 5 above)
Comment (clokep) [reply tests/replication/test_multi_media_repo.py:119]: Yes.

**Pull request merged**